### PR TITLE
Add enduser processor

### PR DESCRIPTION
**Pull request by pavolloffay**
Add end user processor that parses span attributes and adds end user info (id, session, role, scope) as new span attributes.

### CODE REVIEW

File: processors/enduserprocessor/enduserprocessor.go [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+	attribMap := span.Attributes()
+
+	for _, condition := range conditions {
+		attrib, ok := attribMap.Get(condition.Key)
+		if !ok {
+			return false
+		}
+
+		value := attrib.StringVal()
+		// TODO is it correct? it was `value == nil` before
```
Comment (pavolloffay) [inline processors/enduserprocessor/enduserprocessor.go]: @davexroth can you please check this?

Comment (davexroth) [reply processors/enduserprocessor/enduserprocessor.go]: Yep, this is correct.  In the oc collector GetStringValue() returned a pointer.

File: processors/enduserprocessor/hash/hash.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package hash
```
Comment (davexroth) [inline processors/enduserprocessor/hash/hash.go:1]: we may want to move this into the piifilter when we add sessionzation back

File: processors/enduserprocessor/enduserprocessor.go [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+		return user
+	} else {
+		p.logger.Info("Authorization header must be basic or bearer", zap.String("value", value))
+	}
+
+	return nil
+}
+
+func (p *processor) jsonCapture(enduser config, value string) *user {
+	v := p.getJSON(value)
```
Comment (davexroth) [inline processors/enduserprocessor/enduserprocessor.go:214]: this is where it would be great to re-use the previously parsed json

Comment (pavolloffay) [reply processors/enduserprocessor/enduserprocessor.go:214]: But this field might not be reducted right?

File: processors/enduserprocessor/enduserprocessor.go [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+const (
+	enduserIDAttribute      = "enduser.id"
+	enduserRoleAttribute    = "enduser.role"
+	enduserScopeAttribute   = "enduser.scope"
+	enduserSessionAttribute = "session.id"
+)
+
+type config struct {
+	EndUser
+	hashAlgorithm hash.Algorithm
```
Comment (jcchavezs) [inline processors/enduserprocessor/enduserprocessor.go:33]: Not sure this is still a thing maybe @davexroth can confirm? In PII filter we only use hash.

File: processors/enduserprocessor/enduserprocessor.go [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+		ilss := rs.InstrumentationLibrarySpans()
+		for j := 0; j < ilss.Len(); j++ {
+			ils := ilss.At(j)
+
+			spans := ils.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+
+				span.Attributes().ForEach(func(k string, v pdata.AttributeValue) {
+					unindexedKey := piifilterprocessor.UnindexedKey(k)
```
Comment (jcchavezs) [inline processors/enduserprocessor/enduserprocessor.go:64]: I know this is ported from the previous collector but let's avoid leaking this kind of logic. The code behind this function is simple enough to copy into this processor, also it is not exclusive for PII filter.

File: processors/enduserprocessor/enduserprocessor.go [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+	case "authheader":
+		u = p.authHeaderCapture(enduser, value)
+	case "json":
+		u = p.jsonCapture(enduser, value)
+	case "urlencoded":
+		u = p.urlencodedCapture(enduser, value)
+	case "cookie":
+		u = p.cookieCapture(enduser, value)
+	default:
+		p.logger.Warn("Unknown enduser type, skipping", zap.String("type", enduser.Type), zap.String("key", enduser.AttributeKey))
```
Comment (jcchavezs) [inline processors/enduserprocessor/enduserprocessor.go:100]: I think `type` is a very generic key. I think the idea of having this fields is to be able to filter by them so maybe we use something more specific? cc @davexroth

File: processors/enduserprocessor/enduserprocessor.go [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+}
+
+func (p *processor) scopeCapture(value string) *user {
+	return &user{scope: value}
+}
+
+func (p *processor) authHeaderCapture(enduser config, value string) *user {
+	lcValue := strings.ToLower(value)
+	if strings.HasPrefix(lcValue, "bearer ") {
+		tokenString := value[7:]
```
Comment (jcchavezs) [inline processors/enduserprocessor/enduserprocessor.go:178]:
```suggestion
		tokenString := value[7:] // 7 = len("bearer ")
```

File: processors/enduserprocessor/enduserprocessor.go [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+	return &user{role: value}
+}
+
+func (p *processor) scopeCapture(value string) *user {
+	return &user{scope: value}
+}
+
+func (p *processor) authHeaderCapture(enduser config, value string) *user {
+	lcValue := strings.ToLower(value)
+	if strings.HasPrefix(lcValue, "bearer ") {
```
Comment (jcchavezs) [inline processors/enduserprocessor/enduserprocessor.go:177]: how about encapsulating the logic inside this ifs with meaningful function/methods?

File: processors/enduserprocessor/enduserprocessor.go [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+		}
+		if !matched {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (p *processor) idCapture(value string) *user {
```
Comment (jcchavezs) [inline processors/enduserprocessor/enduserprocessor.go:163]: how about this being a function? no need to use the reference to `p`.

File: processors/enduserprocessor/enduserprocessor.go [hunk 10]
```diff
@@ -0,0 +1,534 @@
...
+	if _, ok := attributes.Get(key); ok {
+		return
+	}
+	attributes.Insert(key, pdata.NewAttributeValueString(value))
+}
+
+// only capture the user info if all conditions
+// are true.  If an conditions key does not exist
+// in the span, that is considered a failed condition
+func (p *processor) passesConditions(span pdata.Span, conditions []Condition) bool {
```
Comment (jcchavezs) [inline processors/enduserprocessor/enduserprocessor.go:136]: how about turning this into a function?

File: processors/enduserprocessor/enduserprocessor.go [hunk 11]
```diff
@@ -0,0 +1,534 @@
...
+		}
+		return u
+	} else if strings.HasPrefix(lcValue, "basic ") {
+		token := value[6:]
+		str, err := base64.StdEncoding.DecodeString(token)
+		if err != nil {
+			p.logger.Info("Could not decode basic token", zap.String("value", value))
+			return nil
+		}
+		creds := bytes.SplitN(str, []byte(":"), 2)
```
Comment (jcchavezs) [inline processors/enduserprocessor/enduserprocessor.go:197]: 👍

**Comment by pavolloffay** [comment]
Moving to other repo. @jcchavezs this PR only migrates code I will not change any functionality at this point.

**Pull request closed without merging**